server: avoid deadlock when initing additional stores

tbg opened the pull request:
We need to start node liveness before waiting for additional store init.

Otherwise, we can end up in a situation where each node is sitting on the
channel and nobody has started their liveness yet. The sender to the channel
will first have to get an Increment through KV, but if nobody acquires the
lease (since nobody's heartbeat loop is running), this will never happen.

In practice, *most of the time*, there is no deadlock because the lease
acquisition path performs a synchronous heartbeat to the own entry in most
cases (ignoring the fact that liveness hasn't been started yet). But there is
also another path where someone else's epoch needs to be incremented, and this
path also checks if the node itself is live - which it won't necessarily be
(liveness loop is not running yet).

Fixes #106706

Epic: None
Release note (bug fix): a rare (!) situation in which nodes would get stuck
during start-up was addressed. This is unlikely to have been encountered by
production users This is unlikely to have been encountered by users. If so, it
would manifest itself through a stack frame sitting on a select in
`waitForAdditionalStoreInit` for extended periods of time (i.e. minutes).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107124)
<!-- Reviewable:end -->

tbg committed:
server: start node liveness before waiting for additional store init

Otherwise, we can end up in a situation where each node is sitting on the
channel and nobody has started their liveness yet. The sender to the channel
will first have to get an Increment through KV, but if nobody acquires the
lease (since nobody's heartbeat loop is running), this will never happen.

In practice, *most of the time*, there is no deadlock because the lease
acquisition path performs a synchronous heartbeat to the own entry in most
cases (ignoring the fact that liveness hasn't been started yet). But there is
also another path where someone else's epoch needs to be incremented, and this
path also checks if the node itself is live - which it won't necessarily be
(liveness loop is not running yet).

Fixes #106706

Epic: None
Release note (bug fix): a rare (!) situation in which nodes would get stuck
during start-up was addressed. This is unlikely to have been encountered by
production users This is unlikely to have been encountered by users. If so, it
would manifest itself through a stack frame sitting on a select in
`waitForAdditionalStoreInit` for extended periods of time (i.e. minutes).

tbg committed:
liveness: error on sync heartbeats until liveness started

If we rely on sync heartbeats, there's an issue. They were very effective in
hiding the problem in #106706 so at least in our testing, allow sync heartbeats
only once there are also async heartbeats.

Epic: none
Release note: None

erikgrinaker commented on pkg/server/server.go:
```diff
@@ -1900,13 +1900,20 @@ func (s *Server) PreStart(ctx context.Context) error {
 	// - we'll need to do it after having opened up the RPC floodgates (due to
 	//   the hazard described in Node.start, around initializing additional
 	//   stores)
+	// - we'll need to do it after starting node liveness, see:
```
nit: update "two reasons" to "three reason" above (or just "several reasons".

erikgrinaker commented on pkg/server/server.go:
```diff
@@ -1900,13 +1900,20 @@ func (s *Server) PreStart(ctx context.Context) error {
 	// initialized.
 	s.grpc.setMode(modeOperational)
 
+	// Begin the node liveness heartbeat. Add a callback which records the local
+	// store "last up" timestamp for every store whenever the liveness record is
+	// updated.
```
I think the comment about the callback is stale here.

erikgrinaker reviewed (approved):
We don't need a release note, this was broken recently in #103601.

LGTM once comments on #107265 are addressed. Thanks!

erikgrinaker commented:
> We don't need a release note, this was broken recently in #103601.

Nah, I might be wrong about that. I thought we moved the startup there, but looks like we didn't.

tbg commented:
TFTR!

bors r=erikgrinaker
